src/libopensc/card-dtrust.c

```diff
@@ -0,0 +1,456 @@
+/*
+ * card-dtrust.c: Support for (CardOS based) D-Trust Signature Cards
+ *
+ * Copyright (C) 2023 Mario Haustein <mario.haustein@hrz.tu-chemnitz.de>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ *
+ * based on card-cardos.c
+ */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <stdlib.h>
+#include <string.h>
+#include "internal.h"
+#include "asn1.h"
+
+static const struct sc_card_operations *iso_ops = NULL;
+
+static struct sc_card_operations dtrust_ops;
+static struct sc_card_driver dtrust_drv =
+{
+	"D-Trust Signature Card",
+	"dtrust",
+	&dtrust_ops,
+	NULL, 0, NULL
+};
+
+/* internal structure to save the current security environment */
+struct dtrust_drv_data_t
+{
+	const sc_security_env_t *env;
+};
+
+static const struct sc_atr_table dtrust_atrs[] =
+{
+	/* D-Trust Signature Card v4.1 and v4.4 - CardOS 5.4
+	 *
+	 * The ATR was intentionally omitted from minidriver_registration[] within win32/customactions.cpp
+	 * as it is identical to that of CardOS v5.4 and therefore already included.
+	 * Any new ATR may need an entry in minidriver_registration[]. */
+	{ "3b:d2:18:00:81:31:fe:58:c9:04:11", NULL, NULL, SC_CARD_TYPE_DTRUST_V4_1_STD, 0, NULL },
+	{ NULL,                               NULL, NULL, 0,                            0, NULL }
+};
+
+static struct dtrust_supported_ec_curves
+{
+	struct sc_object_id oid;
+	size_t size;
+} dtrust_curves[] =
+{
+	{ .oid = {{ 1, 2, 840, 10045, 3, 1, 7, -1 }}, .size = 256 },	/* secp256r1 */
+	{ .oid = {{ -1 }},                            .size =   0 },
+};
+
+static int _dtrust_match_cardos(sc_card_t *card)
+{
+	int r;
+	u8 buf[32];
+
+	/* check OS version */
+	r = sc_get_data(card, 0x0182, buf, 32);
+	LOG_TEST_RET(card->ctx, r, "OS version check failed");
+
+	if(r != 2 || buf[0] != 0xc9 || buf[1] != 0x04)
+		return SC_ERROR_WRONG_CARD;
+
+	/* check product name */
+	r = sc_get_data(card, 0x0180, buf, 32);
+	LOG_TEST_RET(card->ctx, r, "Product name check failed");
+
+	if(strcmp((char*)buf, "CardOS V5.4     2019"))
```

Comment:
either check the `r > strlen("CardOS V5.4     2019")` before calling strcmp (and use strncmp), initialze the `buf` or even better call memcmp to avoid issues if the data will not be null terminated.

Reply:
I amended the source such that `r` will be explicitly checked and `memcmp()` is used instead of `strcmp()`.